### PR TITLE
fix: Fix SSR & AmplifyOutput types when using Gen2 configuration files

### PR DESCRIPTION
**Pull request by jimblanc**
#### Description of changes

This change updated the `AmplifyOutputs` type & SSR utilities to fully support the Gen2 configuration file format.

Changes:
- Loosen up types on the `model_introspection` field in `AmplifyOutputs`
  - This is required as TypeScript is not able to properly infer `ModelIntrospectionSchema` from the JSON file. This is only required for configuration file parsing, ultimately the type present in `ResourcesConfig` (which the library uses internally after config file parsing) will be `ModelIntrospectionSchema`. For more context see: https://github.com/aws-amplify/amplify-js/pull/13234
- Add support for `AmplifyOutputs` to SSR utilities to enable server-side API usage against a Gen2 back-end
- Add support for `AmplifyOutputs` to the Amplify singleton
  - **Important context:** This change is the root cause of the apparent bundle size increases observed in this PR. Previously, support for `AmplifyOutputs` was only added to `initSingleton` (https://github.com/aws-amplify/amplify-js/pull/13150), which is the primary method by which customers access the singleton. As our APIs interact with the `Amplify` singleton directly (to retrieve configuration, for example), the bundle size impact of adding Gen2 parsing to `initSingleton` (which customers would almost always use in v6) was obfuscated from the bundle size tests. The changes to the bundle limits in this PR simply bring the tests in-line with what real customers would see in practice as a result of the Gen2 changes.
- Standardized config file parsing across the library
  - Previously we were doing this individually in three different places (i.e. `initSingleton`, `Amplify`, and `adapter-nextjs`), standardized all three places to use the same config parsing utility.

#### Description of how you validated changes
- Tested Gen2 config file parsing in a sample app with both auth & data resources configured
- Smoke tested a handful of SSR APIs using a Gen2 configuration file

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/adapter-nextjs/src/utils/getAmplifyConfig.ts
Comment (jimblanc) [inline packages/adapter-nextjs/src/utils/getAmplifyConfig.ts:1]: This SSR-specific utility has been replaced with the universal `parseAmplifyConfig` utility.

File: packages/aws-amplify/package.json
Comment (jimblanc) [inline packages/aws-amplify/package.json:1]: See PR description for context on apparent bundle size increases. The delta in these increases line up pretty well with the size of the new Gen2 parser.

File: packages/core/src/singleton/Amplify.ts [hunk 1]
```diff
@@ -48,16 +49,10 @@ export class AmplifyClass {
 	 * @param libraryOptions - Additional options for customizing the behavior of the library.
 	 */
 	configure(
-		resourcesConfig: ResourcesConfig | LegacyConfig,
+		resourcesConfig: ResourcesConfig | LegacyConfig | AmplifyOutputs,
 		libraryOptions?: LibraryOptions,
 	): void {
-		let resolvedResourceConfig: ResourcesConfig;
```
Comment (jimblanc) [inline packages/core/src/singleton/Amplify.ts:54]: Amplify singleton did not previously support Gen2 configuration files.

File: packages/core/__tests__/utils/parseAmplifyConfig.test.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import { ResourcesConfig } from '../../src';
+import { parseAmplifyConfig } from '../../src/libraryUtils';
+import { parseAWSExports } from '../../src/parseAWSExports';
+import { isAmplifyOutputs, parseAmplifyOutputs } from '../../src/parseAmplifyOutputs';
+
+jest.mock('../../src/parseAWSExports');
+jest.mock('../../src/parseAmplifyOutputs');
+
+const testAmplifyOutputs = {
```
Comment (Samaritan1011001) [inline packages/core/__tests__/utils/parseAmplifyConfig.test.ts:9]: Do we want to test the parsing of the complete config given that we check for each block in the `parseAmplifyOutputs` function?

**Review by Samaritan1011001 — APPROVED** [review]
thank you!

File: packages/adapter-nextjs/src/utils/getAmplifyConfig.ts
Comment (ashika112) [reply packages/adapter-nextjs/src/utils/getAmplifyConfig.ts:1]: Wondering if there is anything more to why this was not moved to an utility and use across in first place?

File: packages/core/__tests__/utils/parseAmplifyConfig.test.ts (hunk 2 above)
Comment (jimblanc) [reply packages/core/__tests__/utils/parseAmplifyConfig.test.ts:9]: IMO we don't need to here considering each parse function has its own tests. This test is simply ensuring that the correct parser is selected for a given config shape

File: packages/core/src/parseAmplifyOutputs.ts [hunk 3]
```diff
@@ -215,7 +216,7 @@ function parseData(
 		defaultAuthMode: getGraphQLAuthMode(default_authorization_type),
 		region: aws_region,
 		apiKey: api_key,
-		modelIntrospection: model_introspection,
+		modelIntrospection: model_introspection as ModelIntrospectionSchema,
```
[the remark this answers is not in the document]
Comment (jimblanc) [reply packages/core/src/parseAmplifyOutputs.ts:219]: If we're not able to infer the type correctly from the JSON, we don't really have a choice IMO. This is similar to what we're doing already in `parseAWSExports` (expect there it's `any`). We can't lean on TypeScript alone to validate config structure regardless for our non-TS customers, e.g. how we make runtime assertions about configuration state elsewhere in the library.

File: packages/adapter-nextjs/src/utils/getAmplifyConfig.ts
Comment (jimblanc) [reply packages/adapter-nextjs/src/utils/getAmplifyConfig.ts:1]: Checked with Hui, originally we were reading the configuration in a different way for SSR, but that changed before launch. No reason to not unify at this time as far as I can tell.

**Review by HuiSF — COMMENTED** [review]
Thanks for consolidating the parser function, it looks much cleaner now. 

One thing we've might missed is that we are currently exporting the original `parseAmplifyConfig` function from the path `aws-amplify/utils`, to support the generic SSR adapter [use cases](https://docs.amplify.aws/react/build-a-backend/server-side-rendering/nuxt/#implement-01amplify-apisserverts). 

We may need to update the function being exported here:

https://github.com/aws-amplify/amplify-js/blob/1625424a5f04ed6eb2ad9386ffc31d8e25f88e90/packages/aws-amplify/src/utils/index.ts#L21C10-L21C25

File: packages/core/__tests__/utils/parseAmplifyConfig.test.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	}
+};
+
+describe('parseAmplifyConfig', () => {
+	const mockParseAWSExports = parseAWSExports as jest.Mock;
+	const mockParseAmplifyOutputs = parseAmplifyOutputs as jest.Mock;
+	const mockIsAmplifyOutputs = isAmplifyOutputs as unknown as jest.Mock;
+
+	beforeEach(() => {
+		jest.clearAllMocks();
```
[the remark this answers is not in the document]
Comment (jimblanc) [reply packages/core/__tests__/utils/parseAmplifyConfig.test.ts:40]: I'd prefer to keep this in place to be able to make assertions about how many times something is called.

**Comment by jimblanc** [comment]
> Thanks for consolidating the parser function, it looks much cleaner now.
> 
> One thing we've might missed is that we are currently exporting the original `parseAmplifyConfig` function from the path `aws-amplify/utils`, to support the generic SSR adapter [use cases](https://docs.amplify.aws/react/build-a-backend/server-side-rendering/nuxt/#implement-01amplify-apisserverts).
> 
> We may need to update the function being exported [here](https://github.com/aws-amplify/amplify-js/blob/1625424a5f04ed6eb2ad9386ffc31d8e25f88e90/packages/aws-amplify/src/utils/index.ts#L21C10-L21C25).

Good catch with the generic SSR adapter use-case, updated to export `parseAmplifyConfig` directly. While the type surface is technically different, it should still work for anyone passing in a config object as described in the docs.

**Review by HuiSF — APPROVED** [review]

**Review by ashika112 — APPROVED** [review]

**Review by Samaritan1011001 — APPROVED** [review]

**Pull request merged**